replica/tablet_sstable_set.hh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+#include <absl/container/flat_hash_map.h>
+
+#include "compaction_group.hh"
+#include "locator/tablets.hh"
+#include "sstables/sstable_set.hh"
+
+namespace replica {
+
+using namespace locator;
+
+// This sstable set provides access to all the stables in the table, using a snapshot of all
+// its tablets/storage_groups compound_sstable_set:s.
+// The managed sets cannot be modified through tablet_sstable_set, but only jointly read from, so insert() and erase() are disabled.
+class tablet_sstable_set : public sstables::sstable_set_impl {
+    schema_ptr _schema;
+    locator::tablet_map _tablet_map;
+    // Keep a single (compound) sstable_set per tablet/storage_group
+    absl::flat_hash_map<size_t, lw_shared_ptr<const sstables::sstable_set>, absl::Hash<size_t>> _sstable_sets;
+    // Used when ordering is required for correctness, but hot paths will use flat_hash_map
+    // which provides faster lookup time.
+    std::set<size_t> _sstable_set_ids;
+    size_t _size = 0;
+    uint64_t _bytes_on_disk = 0;
+
+public:
+    tablet_sstable_set(schema_ptr s, const storage_group_manager& sgm, const locator::tablet_map& tmap)
+        : _schema(std::move(s))
+        , _tablet_map(tmap.tablet_count())
+    {
+        sgm.for_each_storage_group([this] (size_t id, storage_group& sg) {
+            auto set = sg.make_sstable_set();
+            _size += set->size();
+            _bytes_on_disk += set->bytes_on_disk();
+            _sstable_sets[id] = std::move(set);
+            _sstable_set_ids.insert(id);
+        });
+    }
+
+    tablet_sstable_set(const tablet_sstable_set& o)
+        : _schema(o._schema)
+        , _tablet_map(o._tablet_map.tablet_count())
+        , _sstable_sets(o._sstable_sets)
+        , _sstable_set_ids(o._sstable_set_ids)
+        , _size(o._size)
+        , _bytes_on_disk(o._bytes_on_disk)
+    {}
```

Comment:
are we able to switch to the default copy ctor here? I think it should be possible. tablet map can be copied, and will have same effect as invoking ctor with tablet count of original map.

Reply:
yes the default ctor worked.